File: python/SelfProgenyImputer.py
from __future__ import annotations

# coding: utf-8
# SelfProgenyImputer.py
# 親が自殖で親がimputedなとき後代をimputeする

from math import log
from typing import List, Tuple

from GenoRecord import GenoRecord
from VCFHMM import *
from Genotype import Genotype


#################### SelfProgenyImputer ####################

MIN_PROB = -1e300

class SelfProgenyImputer(VCFHMM[GenoRecord]):
	DP = List[Tuple[float, int]]	# (log of probability, prev h)
	
	def __init__(self, records: list[GenoRecord], map_: Map, w: float) -> None:
		VCFHMM.__init__(self, records, map_)
		self.records: list[GenoRecord] = records
		# crossover values
		# 後代
		self.Cc = [ Map.Kosambi(self.dist(r1, r2))
							for r1, r2 in zip(self.records, self.records[1:]) ]
	
	def num_progenies(self) -> int:
		return len(self.records[0].geno) - 1
	
	def progeny_genotype(self, h: int, parent_gt: int) -> int:
		hc2, hc1 = divmod(h, 2)
		return ((parent_gt >> hc1) & 1) | (((parent_gt >> hc2) & 1) << 1)
	
	# i: record index
	def emission_probability(self, h: int, parent_gt: int, oc: int) -> float:
		gt_prog = self.progeny_genotype(h, parent_gt)
		return self.E[gt_prog][oc&3]
	
	def initialize_dp(self, iprog: int) -> list[DP]:
		M = len(self.records)	# マーカー数
		dp = [ [ (MIN_PROB, 0) ] * 4 for _ in range(M) ]
		record = self.records[0]
		parent_gt = self.records[0].geno[0]
		# observed progeny
		oc = record.geno[iprog+1]
		for h in range(4):		# hidden state
			E_all = self.emission_probability(h, parent_gt, oc)
			dp[0][h] = (E_all, h)
		return dp
	
	def transition_probability(self, h: int, prev_h: int, cc: float) -> float:
		h2, h1 = divmod(h, 2)
		prev_h2, prev_h1 = divmod(prev_h, 2)
		# 遷移確率 0なら(1-c)、1ならcを掛ける
		return (log(cc if h1 != prev_h1 else 1.0 - cc) +
				log(cc if h2 != prev_h2 else 1.0 - cc))
	
	def update_dp(self, i: int, iprog: int, dp: list[DP]) -> None:
		record = self.records[i]
		cc = self.Cc[i-1]	# 遷移確率
		parent_gt = self.records[i].geno[0]
		# observed progs
		oc = record.unphased(iprog+1)
		
		for h in range(4):		# hidden state
			E_all = self.emission_probability(h, parent_gt, oc)
			
			for prev_h in range(4):
				T_all = self.transition_probability(h, prev_h, cc)
				
				prob = dp[i-1][prev_h][0] + (T_all + E_all)
				dp[i][h] = max(dp[i][h], (prob, prev_h))
	
	def update_genotypes(self, hs: list[int], iprog: int) -> None:
		M = len(self.records)
		N = self.num_progenies()
		for i in range(M):
			record = self.records[i]
			parent_gt = record.geno[0]
			prog_gt = self.progeny_genotype(hs[i], parent_gt)
			record.geno[iprog+1] = prog_gt | 4
	
	def impute(self, iprog: int) -> None:
		# DP
		# 外側から、マーカー、ハプロタイプの状態
		M = len(self.records)
		dp = self.initialize_dp(iprog)
		for i in range(1, M):
			self.update_dp(i, iprog, dp)
		
		hs = self.trace_back(dp)
		self.update_genotypes(hs, iprog)

File: src/SelfProgenyImputer.cpp
#include <cmath>
#include "../include/SelfProgenyImputer.h"
#include "../include/common.h"

using namespace std;

SelfProgenyImputer::SelfProgenyImputer(const vector<GenoRecord *>& rs,
									 size_t iprog, const Map& map_, double w) :
					VCFHMM(rs, map_, w), records(rs), ic(iprog), Cc(calc_Cc(rs))
					{ }

vector<double> SelfProgenyImputer::calc_Cc(
							const vector<GenoRecord *>& rs) const {
	const size_t	M = rs.size();
	vector<double>	Cc(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cc[i] = Map::Kosambi(dist(rs[i], rs[i+1]));
	}
	return Cc;
}

double SelfProgenyImputer::emission_probability(int h, int parent_gt,
																int oc) const {
	const int	gt_prog = progeny_genotype(h, parent_gt);
	return E[gt_prog][oc];
}

double SelfProgenyImputer::transition_probability(size_t i,
													int prev_h, int h) const {
	const double	cc = Cc[i-1];
	const int	d = h ^ prev_h;
	return (log((d & 1) == 1 ? cc : 1.0 - cc) +
			log((d >> 1) == 1 ? cc : 1.0 - cc));
}

vector<SelfProgenyImputer::DP> SelfProgenyImputer::initialize_dp() const {
	vector<DP>	dp(M(), DP(4, pair<double, int>(MIN_PROB, 0)));
	const GenoRecord	*record = records[0];
	const int	parent_gt = record->get_geno()[0];
	const int	oc = record->get_geno()[ic+1];
	for(int h = 0; h < 4; ++h) {
		const double	E_all = emission_probability(h, parent_gt, oc);
		dp[0][h] = make_pair(E_all, h);
	}
	return dp;
}

void SelfProgenyImputer::update_dp(size_t i, vector<DP>& dp) const {
	const GenoRecord	*record = records[i];
	const int	parent_gt = record->get_geno()[0];
	const int	oc = record->get_geno()[ic+1];
	for(int h = 0; h < 4; ++h) {
		const double	E_all = emission_probability(h, parent_gt, oc);
		
		for(int prev_h = 0; prev_h < 4; ++prev_h) {
			const double	Tp = transition_probability(i, prev_h, h);
			const double	prob = dp[i-1][prev_h].first + (Tp + E_all);
			dp[i][h] = max(dp[i][h], make_pair(prob, prev_h));
		}
	}
}

void SelfProgenyImputer::update_genotypes(const vector<int>& hs) {
	for(size_t i = 0; i < this->M(); ++i) {
		GenoRecord	*record = records[i];
		const int	parent_gt = record->get_geno()[0];
		const int	prog_gt = progeny_genotype(hs[i], parent_gt);
		record->set_geno(ic + 1, prog_gt | 4);
	}
}

void SelfProgenyImputer::impute() {
	// DP
	vector<DP>	dp = initialize_dp();
	for(int i = 1; i < (int)this->M(); ++i) {
		update_dp(i, dp);
	}
	
	const vector<int>	hs = trace_back(dp);
	update_genotypes(hs);
}
